pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -409,55 +409,57 @@ func (p *PoolManager) AddMacToWaitingConfig(allocations map[string]string, paren
 }
 
 // Remove all the mac addresses from the waiting configmap this mean the vm was saved in the etcd and pass validations
-func (p *PoolManager) MarkVMAsReady(vm *kubevirt.VirtualMachine, parentLogger logr.Logger) error {
+func (p *PoolManager) MarkVMAsReady(vm *kubevirt.VirtualMachine, latestPersistedTransactionTimeStamp string, parentLogger logr.Logger) error {
 	logger := parentLogger.WithName("MarkVMAsReady")
 
 	p.poolMutex.Lock()
 	defer p.poolMutex.Unlock()
+	vmFullName := VmNamespaced(vm)
+	var updatedMacsList []string
 
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		// refresh ConfigMaps instance
-		configMap, err := p.kubeClient.CoreV1().ConfigMaps(p.managerNamespace).Get(context.TODO(), names.WAITING_VMS_CONFIGMAP, metav1.GetOptions{})
-		if err != nil {
-			return errors.Wrap(err, "Failed to refresh manager's configmap instance")
-		}
-
-		if configMap.Data == nil {
-			logger.Info("the configMap is empty")
-			return nil
-		}
-
-		if len(vm.Spec.Template.Spec.Domain.Devices.Interfaces) == 0 {
-			logger.Info("interface list is empty")
-			return nil
-		}
+	vmMacMap, err := p.getInstanceMacMap(vmFullName)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to get VmMacMap for vm %s", vmFullName)
+	}
 
-		logger.V(1).Info("set vm's mac to status allocated", "vm interfaces", vm.Spec.Template.Spec.Domain.Devices.Interfaces)
-		for _, vmInterface := range vm.Spec.Template.Spec.Domain.Devices.Interfaces {
-			if vmInterface.MacAddress != "" {
-				if _, exist := p.macPoolMap[vmInterface.MacAddress]; exist {
-					p.macPoolMap[vmInterface.MacAddress] = AllocationStatusAllocated
-				}
-				macAddress := strings.Replace(vmInterface.MacAddress, ":", "-", 5)
-				delete(configMap.Data, macAddress)
+	vmPersistedInterfaceList := getVirtualMachineInterfaces(vm)
+	logger.V(1).Info("checking macMap Alignment", "vmMacMap", vmMacMap, "interfaces", vmPersistedInterfaceList, "latestPersistedTransactionTimeStamp", latestPersistedTransactionTimeStamp)
+	for macAddress, macEntry := range vmMacMap {
+		logger.V(1).Info("macAddress params:", "interfaceName", macEntry.macInstanceKey, "transactionTimeStamp", macEntry.transactionTimestamp)
+		if p.isMacUpdateRequired(macAddress) {
+			macReadyForUpdate, err := p.isMacReadyForTransactionUpdate(macAddress, latestPersistedTransactionTimeStamp)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to check mac entry readiness")
+			}
+			if macReadyForUpdate {
+				logger.V(1).Info("macAddress ready for update")
+				p.alignMacEntryAccordingToVmInterface(macAddress, macEntry, vmPersistedInterfaceList)
```

Comment:
We can end up deleting the macEntry here from the map, that means that we are changing the map in the middle of a for loop, is this dangerous ?

Reply:
I don't think so, but I think we should refactor this a bit anyway